### DISCUSSION TITLE
MinIO API endpoint prompting for client certificate before redirecting to console

### DISCUSSION THREAD

**Discussion by mbklein** [minio/minio#13371] [Q&A] [answered]
My team has been using the `minio/minio` image in development to build our S3-backed application for years, but this issue didn’t come up until recently. Here’s how our minio service is configured:

- Docker volume bound to `/data`
- Trusted certificate and key bound to `/root/.minio/certs/public.crt` and `/root/.minio/certs/public.key`
- Server started with `--address devbox.our.domain.example.edu:9000 --console-address devbox.our.domain.example.edu:19000 /data`
- Ports 9000 and 19000 forwarded to the host

This all works just fine using the image `minio/minio:RELEASE.2021-08-25T00-41-18Z`. But if I try to use `minio/minio:latest` (digest `f67e464d190f`, pushed on 2021-10-03), the first connection from a browser to `https://devbox.our.domain.example.edu:9000/` triggers a popup window asking for the browser to supply a client certificate. This happens whether the connection is made to the root (which should just be forwarding to the console on port 19000), or during a presigned `POST` request from our app.

If the user hits cancel, the request goes through as expected, everything seems to work fine, and the prompt doesn’t come up again for some unknown period of time. I haven’t been able to force it by recreating the container and persistent volume, restarting my browser, clearing my cache, or using private browsing mode, but then it will happen again when I’m not expecting it.

When I `curl -v https://devbox.our.domain.example.edu:9000/` using the older docker image, I see:

```
* TLSv1.2 (OUT), TLS handshake, Client hello (1):
* TLSv1.2 (IN), TLS handshake, Server hello (2):
* TLSv1.2 (IN), TLS handshake, Certificate (11):
* TLSv1.2 (IN), TLS handshake, Server key exchange (12):
* TLSv1.2 (IN), TLS handshake, Server finished (14):
* TLSv1.2 (OUT), TLS handshake, Client key exchange (16):
* TLSv1.2 (OUT), TLS change cipher, Change cipher spec (1):
* TLSv1.2 (OUT), TLS handshake, Finished (20):
* TLSv1.2 (IN), TLS change cipher, Change cipher spec (1):
* TLSv1.2 (IN), TLS handshake, Finished (20):
* SSL connection using TLSv1.2 / ECDHE-RSA-CHACHA20-POLY1305
* ALPN, server accepted to use http/1.1
* Server certificate:
* [...certificate details...]
```

But the same request on the latest docker image shows:

```
* TLSv1.2 (OUT), TLS handshake, Client hello (1):
* TLSv1.2 (IN), TLS handshake, Server hello (2):
* TLSv1.2 (IN), TLS handshake, Certificate (11):
* TLSv1.2 (IN), TLS handshake, Server key exchange (12):
* TLSv1.2 (IN), TLS handshake, Request CERT (13):  <-- This line is new
* TLSv1.2 (IN), TLS handshake, Server finished (14):
* TLSv1.2 (OUT), TLS handshake, Certificate (11):  <-- This line is also new
* TLSv1.2 (OUT), TLS handshake, Client key exchange (16):
* TLSv1.2 (OUT), TLS change cipher, Change cipher spec (1):
* TLSv1.2 (OUT), TLS handshake, Finished (20):
* TLSv1.2 (IN), TLS change cipher, Change cipher spec (1):
* TLSv1.2 (IN), TLS handshake, Finished (20):
* SSL connection using TLSv1.2 / ECDHE-RSA-AES128-GCM-SHA256
* ALPN, server accepted to use http/1.1
* Server certificate:
* [...certificate details...]
```

What changed between the `2021-08-25T00-41-18Z` release and `latest` that might explain this, and is there a way to turn off the client cert request while still using TLS on the server end?

**Comment by harshavardhana** [comment, marked as answer]
> What changed between the `2021-08-25T00-41-18Z` release and `latest` that might explain this, and is there a way to turn off the client cert request while still using TLS on the server end?

This is basically the TLS Client cert issue perhaps @aead

**Reply by mbklein** [reply]
Are you saying that this is a manifestation of a known issue that has not yet been resolved? Or is there something I can configure to prevent the server from requesting a client cert?

**Reply by aead** [reply]
This (window popup asking for a certificate) seems to be a sideeffect of MinIO STS API `AssumeRoleWithCertificate`.
MinIO asks the client to provide a certificate for mTLS (and issuing temp. S3 credentials). A client may choose to not send a certificate. It is only needed when the client hits the AssumeRoleWithCertificate` STS API. For any other API operation (i.e. all S3 API operations) no client certificate is needed / should be provided.

Usually, client applications should simply not send a client certificate if they have none.
Currently, you cannot configure MinIO to not request a client certificate.

However, we could not ask for a client certificate if `MINIO_IDENTITY_TLS_ENABLE=off`. So, MinIO would only ask for client certificates if `MINIO_IDENTITY_TLS_ENABLE=on`. @harshavardhana

**Reply by aead** [reply]
Maybe like this:
```
diff --git a/internal/http/server.go b/internal/http/server.go
index ad953d2ea..080861981 100644
--- a/internal/http/server.go
+++ b/internal/http/server.go
@@ -28,7 +28,6 @@ import (
    "time"
 
    humanize "github.com/dustin/go-humanize"
-
    "github.com/minio/minio-go/v7/pkg/set"
    "github.com/minio/minio/internal/config"
    "github.com/minio/minio/internal/config/api"
@@ -172,7 +171,9 @@ func NewServer(addrs []string, handler http.Handler, getCert certs.GetCertificat
            MinVersion:               tls.VersionTLS12,
            NextProtos:               []string{"http/1.1", "h2"},
            GetCertificate:           getCert,
-           ClientAuth:               tls.RequestClientCert,
+       }
+       if env.Get("MINIO_IDENTITY_TLS_ENABLE", config.EnableOn) == config.EnableOn {
+           tlsConfig.ClientAuth = tls.RequestClientCert
        }
        if secureCiphers || fips.Enabled {
            tlsConfig.CipherSuites = fips.CipherSuitesTLS()
```

**Reply by mbklein** [reply]
Our applications that use the S3 API are handling things just fine. It’s only the interactive users who are getting the popup (when POSTing to the presigned URL our app generates), and while I can just document the fact that they can hit cancel and continue, I’d prefer to provide them a more seamless experience.

An environment variable that would suppress the clients identity request would be a great solution for us. For now, we will keep using the older image.

**Reply by harshavardhana** [reply]
> -           ClientAuth:               tls.RequestClientCert,

I am able to use TLS certificate API without this check being present @aead do we really need it?

**Reply by harshavardhana** [reply]
> > * ```
> >         ClientAuth:               tls.RequestClientCert,
> >   ```
> 
> I am able to use TLS certificate API without this check being present @aead do we really need it?

Sorry, my bad, tls.RequestClientCert is needed and we have to make it `opt-in` to turn it on instead of turn-off @aead IMO. 

Since the API here requires a custom behavior it makes sense for us to avoid it by default and only enable it when the environment variable is set.

**Reply by aead** [reply]
> Since the API here requires a custom behavior it makes sense for us to avoid it by default and only enable it when the environment variable is set.

@harshavardhana Agree. The `AssumeRoleWithCertificate` API should be opt-in. Only if enabled, the server should request a client certificate to not interfere with browsers.

**Comment by dmytrobabeshko** [comment]
Hello,
We are using MinIO as Object storage for GitLab with 'proxy_mode=false'. In this mode, Gitlab returns HTTP 302 redirect with a pre-signed, time-limited object storage URL.  All our users experience a popup window asking for the browser to supply a client certificate. This didn't happen before AssumeRoleWithCertificate was implemented.
Please add an option to turn this off.
Thank you in advance!

**Comment by bryanyip94** [comment]
Hi,
We are facing exactly the same certificate pop-up issue when calling a pre-signed url. Is there already a fix for this in one of the newer releases? Or do I need to change one of the env variables to disable this?
Thank you in advance!

**Reply by harshavardhana** [reply]
> Hi, We are facing exactly the same certificate pop-up issue when calling a pre-signed url. Is there already a fix for this in one of the newer releases? Or do I need to change one of the env variables to disable this? Thank you in advance!

This has been disabled for quite some time, upgrade your local deployment @bryanyip94
